sdk/dotnet/Pulumi.Automation/DictionaryContentsComparer.cs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2016-2019, Pulumi Corporation
+
+using System.Collections.Generic;
+
+namespace Pulumi.Automation
+{
+    /// Compares two dictionaries for equality by content, as F# maps would.
+    internal sealed class DictionaryContentsComparer<K, V> : IEqualityComparer<IDictionary<K, V>> where K : notnull
+    {
+        private readonly IEqualityComparer<K> _keyComparer;
+        private readonly IEqualityComparer<V> _valueComparer;
+
+        public DictionaryContentsComparer(IEqualityComparer<K> keyComparer, IEqualityComparer<V> valueComparer)
+        {
+            this._keyComparer = keyComparer;
+            this._valueComparer = valueComparer;
+        }
+
+        bool IEqualityComparer<IDictionary<K, V>>.Equals(IDictionary<K, V>? x, IDictionary<K, V>? y)
+        {
+            if (x == null)
+            {
+                return y == null;
+            }
+            if (y == null)
+            {
+                return x == null;
+            }
+            if (x.Count != y.Count)
+            {
+                return false;
+            }
```

Comment:
After you have determined that they are both not null you can do:
```csharp
if (ReferenceEquals(x, y))
    return true;
```

Which will ensure they're not the same ref, since it is a reference type. And then instead of a foreach loop you can knock it out with LINQ / set-comparison:
```csharp
// first ensure keys are the same
if (!x.Keys.ToHashSet().SetEquals(y.Keys))
    return false;
    
// then ensure values are the same
if (x.All(xkvp => this._valueComparer.Equals(xkvp.Value, y[xkvp.Key])))
    return true;
    
return false;

Reply:
I guess you're using a comparer for the keys too so the key comparison would need that comparer in the hashset. So like:
```csharp
var xkeys = x.Keys.ToHashSet(this._keyComparer);
var yKeys = y.Keys.ToHashSet(this._keyComparer);
if (!xKeys.SetEquals(yKeys))
    return false;
```

Reply:
The suggested replacement may not use _keyComparer when deciding SetEquals or when looking things up from y[xkvp.Key]. It will use KeyComparer from x and y. This is subtle but potentially very annoying. I'll leave the original code I think.. there are minor stylistic gains to be had by changing it but at least I'm confident in it. It would be tempting if there was a library method already to do this all thing though..

Reply:
I do think that at the very least you want the `ReferenceEquals(...)` check in all of our comparers for reference types.

Reply:
Yeah that suggestion is good, adding it.